jwst/ami/oifits.py

```diff
@@ -41,44 +42,208 @@ def __init__(self, fringefitter, method="median"):
         self.fringe_amplitudes = np.zeros((self.nslices, self.n_baselines))
         self.closure_phases = np.zeros((self.nslices, self.n_closure_phases))
+        self.t3_amplitudes = np.zeros((self.nslices, self.n_closure_phases))
+        self.q4_phases = np.zeros((self.nslices, self.n_closure_amplitudes))
         self.closure_amplitudes = np.zeros((self.nslices, self.n_closure_amplitudes))
         self.pistons = np.zeros((self.nslices, self.n_holes))
         # model parameters
         self.solns = np.zeros((self.nslices, 44))
 
+        # populate with each integration's observables
         for i, nrmslc in enumerate(self.fringe_fitter.nrm_list):
-            self.fringe_phases[i, :] = np.rad2deg(nrmslc.fringephase)  # FPs in degrees
+            self.fringe_phases[i, :] = nrmslc.fringephase  # FPs in radians
             self.fringe_amplitudes[i, :] = nrmslc.fringeamp
-            self.closure_phases[i, :] = np.rad2deg(nrmslc.redundant_cps)  # CPs in degrees
+            self.closure_phases[i, :] = nrmslc.redundant_cps  # CPs in radians
+            self.t3_amplitudes[i, :] = nrmslc.t3_amplitudes
+            self.q4_phases[i, :] = nrmslc.q4_phases # quad phases in radians
             self.closure_amplitudes[i, :] = nrmslc.redundant_cas
-            self.pistons[i, :] = np.rad2deg(
-                nrmslc.fringepistons
-            )  # segment pistons in degrees
+            self.pistons[i, :] = nrmslc.fringepistons # segment pistons in radians
             self.solns[i, :] = nrmslc.soln
 
         self.fringe_amplitudes_squared = self.fringe_amplitudes ** 2  # squared visibilities
 
+    def rotate_matrix(self, cov_mat, theta):
+        """
+        Rotate a covariance matrix by an angle.
+
+        Parameters
+        ----------
+        cov_mat:
+        theta: float
+            Angle by which to rotate the matrix (radians)
+
+        Returns
+        -------
+
+        """
+        c, s = np.cos(theta), np.sin(theta)
+        R_mat = [[c, -s],
+                 [s, c]]
+        # coordinate rotation from real/imaginary to absolute value/phase (modulus/argument)
+        cv_rotated = np.linalg.multi_dot([np.transpose(R_mat), cov_mat, R_mat])
+        return cv_rotated
+
+
+    def average_observables(self, averfunc):
+        """
+        Calculate covariance matrices between fringe amplitudes/fringe phases, 
+        and between triple product amps/closure phases, and closure amplitudes/quad phases.
+        Convert r, theta (modulus, phase) to x,y. Calculate cov(x,y). Rotate resulting
+        2x2 matrix back to r, theta. Take sqrt of relevant covariance element to be error.
+        This must be done with phases in radians.
+
+        Parameters
+        ----------
+        averfunc
+
+        Returns
+        -------
+        avg_sqv, err_sqv, avg_fa, err_fa, avg_fp, err_fp, avg_cp, err_cp, avg_t3amp, err_t3amp, avg_ca, err_ca, avg_q4phi, err_q4phi, avg_pist, err_pist
```

Comment:
Can we add a little more documentation on the inputs and outputs here?

For the return values, they should be separated out, like input parameters are, so that each one can specify a type and have a description.